tests/convergence/ratesp.cpp

```diff
@@ -0,0 +1,410 @@
+   VectorFunctionCoefficient *gradu=nullptr;
+   VectorFunctionCoefficient *curlU=nullptr;
+
+   ConstantCoefficient one(1.0);
+   ParLinearForm b(fespace);
+   ParBilinearForm a(fespace);
+
+   switch (prob)
+   {
+      case 0:
+         //(grad u_ex, grad v) + (u_ex,v)
+         u = new FunctionCoefficient(u_exact);
+         gradu = new VectorFunctionCoefficient(dim,gradu_exact);
+         b.AddDomainIntegrator(new DomainLFGradIntegrator(*gradu));
+         b.AddDomainIntegrator(new DomainLFIntegrator(*u));
+
+         // (grad u, grad v) + (u,v)
+         a.AddDomainIntegrator(new DiffusionIntegrator(one));
+         a.AddDomainIntegrator(new MassIntegrator(one));
+
+         break;
+      case 1:
+         //(curl u_ex, curl v) + (u_ex,v)
+         U = new VectorFunctionCoefficient(dim,U_exact);
+         curlU = new VectorFunctionCoefficient((dim==3)?dim:1,curlU_exact);
+         b.AddDomainIntegrator(new VectorFEDomainLFCurlIntegrator(*curlU));
+         b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(*U));
+
+         // (curl u, curl v) + (u,v)
+         a.AddDomainIntegrator(new CurlCurlIntegrator(one));
+         a.AddDomainIntegrator(new VectorFEMassIntegrator(one));
+         break;
+
+      case 2:
+         //(div u_ex, div v) + (u_ex,v)
+         U = new VectorFunctionCoefficient(dim,U_exact);
+         divU = new FunctionCoefficient(divU_exact);
+         b.AddDomainIntegrator(new VectorFEDomainLFDivIntegrator(*divU));
+         b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(*U));
+
+         // (div u, div v) + (u,v)
+         a.AddDomainIntegrator(new DivDivIntegrator(one));
+         a.AddDomainIntegrator(new VectorFEMassIntegrator(one));
+         break;
+
+      case 3:
+         u = new FunctionCoefficient(u_exact);
+         f = new FunctionCoefficient(rhs_func);
+         gradu = new VectorFunctionCoefficient(dim,gradu_exact);
+         b.AddDomainIntegrator(new DomainLFIntegrator(*f));
+         b.AddBdrFaceIntegrator(
+      new DGDirichletLFIntegrator(*u, one, sigma, kappa));
+         a.AddDomainIntegrator(new DiffusionIntegrator(one));
+         a.AddInteriorFaceIntegrator(new DGDiffusionIntegrator(one, sigma, kappa));
+         a.AddBdrFaceIntegrator(new DGDiffusionIntegrator(one, sigma, kappa));
+         break;   
+
+      default:
+         break;
+   }
+
+   // 10. Perform successive parallel refinements, compute the L2 error and the
+   //     corresponding rate of convergence
+   Convergence rates;
+   for (int l = 0; l <= pr; l++)
+   {
+      b.Assemble();
+      a.Assemble();
+      a.Finalize();
+
+      HypreParMatrix *A = a.ParallelAssemble();
+      HypreParVector *B = b.ParallelAssemble();
+      HypreParVector *X = x.ParallelProject();
+
+      Solver *prec = NULL;
+      IterativeSolver *solver = NULL;
+      switch (prob)
+      {
+         case 0:
+         case 3:
+            prec = new HypreBoomerAMG(*A);
+            dynamic_cast<HypreBoomerAMG *>(prec)->SetPrintLevel(0);
```

Comment:
Is this better than `HypreBoomer AMG * amg_prec = new ...; prec = amg_prec`? I typically try to avoid casts, but I don't really have a good reason. If this is just as good, it's nice to have fewer LOC!

Reply:
Unfortunately, different solvers (CG/GMRES) and preconditioners (AMG/AMS/ADS) are used for the different runs. I think this way we can avoid code duplication and more LOC.